[NestedTensor] Add support for transposed NestedTensors where ragged_idx > 1 for sum and mean operators

jananisriram opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #131520
* #131519
* #131518
* __->__ #131517

Add support for transposed, non-contiguous `NestedTensor`s, where `ragged_idx > 1`, for the aten operators `sum` and `mean`. This diff enables reducing along the jagged dimension for non-contiguous `NestedTensor`s, transposed between non-batch dimensions as well as between a ragged and a non-batch dimension. For example, users can now reduce a `NestedTensor` of shape `(B, M, *, N)` along `*` or `(B, N, M, *)` along `*`.

Parametrize existing unit tests and add new unit tests verifying the accuracy of implementations on `NestedTensor`s that transpose between 2 non-batch dimensions as well as between a ragged and a non-batch dimension.

Differential Revision: [D59847927](https://our.internmc.facebook.com/intern/diff/D59847927/)

jananisriram committed:
[NestedTensor] Add support for transposed NestedTensors where ragged_idx > 1 for sum and mean operators

Add support for transposed, non-contiguous `NestedTensor`s, where `ragged_idx > 1`, for the aten operators `sum` and `mean`. This diff enables reducing along the jagged dimension for non-contiguous `NestedTensor`s, transposed between non-batch dimensions as well as between a ragged and a non-batch dimension. For example, users can now reduce a `NestedTensor` of shape `(B, M, *, N)` along `*` or `(B, N, M, *)` along `*`.

Parametrize existing unit tests and add new unit tests verifying the accuracy of implementations on `NestedTensor`s that transpose between 2 non-batch dimensions as well as between a ragged and a non-batch dimension.

Differential Revision: [D59847927](https://our.internmc.facebook.com/intern/diff/D59847927/)

[ghstack-poisoned]

davidberard98 commented on torch/nested/_internal/ops.py:
```diff
@@ -897,16 +894,38 @@ def sum_dim_IntList(func, *args, **kwargs):
                 )
             # reduction cases: (ragged)
+            # shift reduction dimension of values backward to outer dimension
+            values_transposed_backward = inp._values.permute(
```
nit: can we use a different name other than "backward" to avoid getting confused with "backward" as in "forward/backward pass"

e.g. maybe "values_standard_layout" or "values_ragged_outer"

jananisriram replied:
yeah that sounds good! changed the `backward` and `forward` variable names

davidberard98 commented on torch/nested/_internal/ops.py:
```diff
@@ -897,16 +894,38 @@ def sum_dim_IntList(func, *args, **kwargs):
+            padded = torch.ops.aten._jagged_to_padded_dense_forward(
+                values_transposed_backward.reshape(
+                    values_transposed_backward.shape[0], -1
+                ),  # _jagged_to_padded_dense_forward requires values to be a 2D tensor
```
... `_jagged_to_padded_dense_forward requires values to be a 2D tensor with the ragged dimension as the 0th dimension`

davidberard98 commented on torch/nested/_internal/ops.py:
```diff
@@ -897,16 +894,38 @@ def sum_dim_IntList(func, *args, **kwargs):
+            )
+
+            # shift reduction dimension of padded tensor forward to original ragged dimension
+            padded_transposed_forward = padded_transposed_forward.permute(
```
cc @jbschlosser: I think the permute could potentially change the striding in a non-obvious way.

I don't think this is a problem (and it's what I'd expect), but wanted to flag it in case it's a concern

davidberard98 reviewed (approved):
Generally lgtm! A few nits and one quick comment for @jbschlosser, otherwise looks good!

jananisriram committed:
Update on "[NestedTensor] Add support for transposed NestedTensors where ragged_idx > 1 for sum and mean operators"


Add support for transposed, non-contiguous `NestedTensor`s, where `ragged_idx > 1`, for the aten operators `sum` and `mean`. This diff enables reducing along the jagged dimension for non-contiguous `NestedTensor`s, transposed between non-batch dimensions as well as between a ragged and a non-batch dimension. For example, users can now reduce a `NestedTensor` of shape `(B, M, *, N)` along `*` or `(B, N, M, *)` along `*`.

Parametrize existing unit tests and add new unit tests verifying the accuracy of implementations on `NestedTensor`s that transpose between 2 non-batch dimensions as well as between a ragged and a non-batch dimension.

Differential Revision: [D59847927](https://our.internmc.facebook.com/intern/diff/D59847927/)

[ghstack-poisoned]

jananisriram committed:
Update on "[NestedTensor] Add support for transposed NestedTensors where ragged_idx > 1 for sum and mean operators"


Add support for transposed, non-contiguous `NestedTensor`s, where `ragged_idx > 1`, for the aten operators `sum` and `mean`. This diff enables reducing along the jagged dimension for non-contiguous `NestedTensor`s, transposed between non-batch dimensions as well as between a ragged and a non-batch dimension. For example, users can now reduce a `NestedTensor` of shape `(B, M, *, N)` along `*` or `(B, N, M, *)` along `*`.

Parametrize existing unit tests and add new unit tests verifying the accuracy of implementations on `NestedTensor`s that transpose between 2 non-batch dimensions as well as between a ragged and a non-batch dimension.

Differential Revision: [D59847927](https://our.internmc.facebook.com/intern/diff/D59847927/)

[ghstack-poisoned]

jananisriram committed:
Update on "[NestedTensor] Add support for transposed NestedTensors where ragged_idx > 1 for sum and mean operators"


Add support for transposed, non-contiguous `NestedTensor`s, where `ragged_idx > 1`, for the aten operators `sum` and `mean`. This diff enables reducing along the jagged dimension for non-contiguous `NestedTensor`s, transposed between non-batch dimensions as well as between a ragged and a non-batch dimension. For example, users can now reduce a `NestedTensor` of shape `(B, M, *, N)` along `*` or `(B, N, M, *)` along `*`.

Parametrize existing unit tests and add new unit tests verifying the accuracy of implementations on `NestedTensor`s that transpose between 2 non-batch dimensions as well as between a ragged and a non-batch dimension.

Differential Revision: [D59847927](https://our.internmc.facebook.com/intern/diff/D59847927/)

[ghstack-poisoned]

jananisriram committed:
Update on "[NestedTensor] Add support for transposed NestedTensors where ragged_idx > 1 for sum and mean operators"


Add support for transposed, non-contiguous `NestedTensor`s, where `ragged_idx > 1`, for the aten operators `sum` and `mean`. This diff enables reducing along the jagged dimension for non-contiguous `NestedTensor`s, transposed between non-batch dimensions as well as between a ragged and a non-batch dimension. For example, users can now reduce a `NestedTensor` of shape `(B, M, *, N)` along `*` or `(B, N, M, *)` along `*`.

Parametrize existing unit tests and add new unit tests verifying the accuracy of implementations on `NestedTensor`s that transpose between 2 non-batch dimensions as well as between a ragged and a non-batch dimension.

Differential Revision: [D59847927](https://our.internmc.facebook.com/intern/diff/D59847927/)

[ghstack-poisoned]

jananisriram commented:
@jananisriram has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D59847927).

jananisriram committed:
Update on "[NestedTensor] Add support for transposed NestedTensors where ragged_idx > 1 for sum and mean operators"


Add support for transposed, non-contiguous `NestedTensor`s, where `ragged_idx > 1`, for the aten operators `sum` and `mean`. This diff enables reducing along the jagged dimension for non-contiguous `NestedTensor`s, transposed between non-batch dimensions as well as between a ragged and a non-batch dimension. For example, users can now reduce a `NestedTensor` of shape `(B, M, *, N)` along `*` or `(B, N, M, *)` along `*`.

Parametrize existing unit tests and add new unit tests verifying the accuracy of implementations on `NestedTensor`s that transpose between 2 non-batch dimensions as well as between a ragged and a non-batch dimension.

Differential Revision: [D59847927](https://our.internmc.facebook.com/intern/diff/D59847927/)

[ghstack-poisoned]

jananisriram commented:
@jananisriram has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D59847927).

jananisriram committed:
Update on "[NestedTensor] Add support for transposed NestedTensors where ragged_idx > 1 for sum and mean operators"


Add support for transposed, non-contiguous `NestedTensor`s, where `ragged_idx > 1`, for the aten operators `sum` and `mean`. This diff enables reducing along the jagged dimension for non-contiguous `NestedTensor`s, transposed between non-batch dimensions as well as between a ragged and a non-batch dimension. For example, users can now reduce a `NestedTensor` of shape `(B, M, *, N)` along `*` or `(B, N, M, *)` along `*`.

Parametrize existing unit tests and add new unit tests verifying the accuracy of implementations on `NestedTensor`s that transpose between 2 non-batch dimensions as well as between a ragged and a non-batch dimension.

Differential Revision: [D59847927](https://our.internmc.facebook.com/intern/diff/D59847927/)

[ghstack-poisoned]
